File: src/main/java/no/fintlabs/provider/event/response/ResponseEventTopicProducer.java
package no.fintlabs.provider.event.response;

import no.fintlabs.adapter.models.event.RequestFintEvent;
import no.fintlabs.adapter.models.event.ResponseFintEvent;
import no.fintlabs.kafka.event.EventProducer;
import no.fintlabs.kafka.event.EventProducerFactory;
import no.fintlabs.kafka.event.EventProducerRecord;
import no.fintlabs.kafka.event.topic.EventTopicNameParameters;
import no.fintlabs.provider.kafka.ProviderTopicService;
import org.springframework.stereotype.Service;

import java.time.Duration;

@Service
public class ResponseEventTopicProducer {

    private final EventProducer<Object> eventProducer;
    private final ProviderTopicService topicService;

    public ResponseEventTopicProducer(EventProducerFactory eventProducerFactory, ProviderTopicService topicService) {
        this.eventProducer = eventProducerFactory.createProducer(Object.class);
        this.topicService = topicService;
    }

    public void sendEvent(ResponseFintEvent responseFintEvent, RequestFintEvent requestFintEvent) {
        EventTopicNameParameters topicNameParameters = EventTopicNameParameters
                .builder()
                .orgId(responseFintEvent.getOrgId())
                .domainContext("fint-core")
                .eventName(createEventName(requestFintEvent))
                .build();

        if (!topicService.topicExists(topicNameParameters))
            topicService.ensureTopic(topicNameParameters, Duration.ofDays(2).toMillis());

        eventProducer.send(
                EventProducerRecord.builder()
                        .topicNameParameters(topicNameParameters)
                        .value(responseFintEvent)
                        .build()
        );
    }

    private String createEventName(RequestFintEvent requestFintEvent) {
        return "%s-%s-%s-%s-response".formatted(
                requestFintEvent.getDomainName(),
                requestFintEvent.getPackageName(),
                requestFintEvent.getResourceName(),
                requestFintEvent.getOperationType().toString().toLowerCase()
        );
    }

}

File: src/main/java/no/fintlabs/provider/event/response/ResponseEventTopicProducer.kt
package no.fintlabs.provider.event.response

import no.fintlabs.adapter.models.event.RequestFintEvent
import no.fintlabs.adapter.models.event.ResponseFintEvent
import no.fintlabs.provider.config.ProducerProperties
import no.novari.kafka.producing.ParameterizedProducerRecord
import no.novari.kafka.producing.ParameterizedTemplateFactory
import no.novari.kafka.topic.EventTopicService
import no.novari.kafka.topic.configuration.EventCleanupFrequency
import no.novari.kafka.topic.configuration.EventTopicConfiguration
import no.novari.kafka.topic.name.EventTopicNameParameters
import no.novari.kafka.topic.name.TopicNamePrefixParameters
import org.springframework.stereotype.Service


@Service
class ResponseEventTopicProducer(
    eventProducerFactory: ParameterizedTemplateFactory,
    private val eventTopicService: EventTopicService,
    private val responseProducerProperties: ProducerProperties
) {

    private val eventProducer = eventProducerFactory.createTemplate(ResponseFintEvent::class.java)

    fun sendEvent(responseFintEvent: ResponseFintEvent, requestFintEvent: RequestFintEvent) {
        val topicNameParameters = requestFintEvent.toTopicNameParameters()

        eventTopicService.createOrModifyTopic(
            topicNameParameters,
            EventTopicConfiguration.stepBuilder()
                .partitions(responseProducerProperties.partitions)
                .retentionTime(responseProducerProperties.retentionTime)
                .cleanupFrequency(EventCleanupFrequency.NORMAL)
                .build()
        )

        eventProducer.send(
            ParameterizedProducerRecord.builder<ResponseFintEvent>()
                .topicNameParameters(topicNameParameters)
                .value(responseFintEvent)
                .build()
        )
    }

    private fun RequestFintEvent.toTopicNameParameters() =
        EventTopicNameParameters
            .builder()
            .topicNamePrefixParameters(
                TopicNamePrefixParameters
                    .stepBuilder()
                    .orgId(orgId.replace(".", "-"))
                    .domainContextApplicationDefault()
                    .build()
            )
            .eventName(toTopicEventName())
            .build()

    private fun RequestFintEvent.toTopicEventName(): String = "${domainName}-${packageName}-response"

}
